File: src/main/java/org/proofit/premiumcalc/calculation/PremiumFactory.java
package org.proofit.premiumcalc.calculation;

import org.proofit.premiumcalc.model.RiskType;

import java.util.EnumMap;


public class PremiumFactory {
    static FirePremium configuredFirePremium = new FirePremium();
    static TheftPremium configuredTheftPremium = new TheftPremium();

    private static final EnumMap<RiskType, Premium> riskToPremiumMapping = getDefaultRiskTypeToPremiumMapping();

    public static Premium getPremium(RiskType riskType) {
        return riskToPremiumMapping.get(riskType);
    }

    protected static EnumMap<RiskType, Premium> getDefaultRiskTypeToPremiumMapping() {
        EnumMap<RiskType, Premium> map = new EnumMap<>(RiskType.class);
        map.put(RiskType.FIRE, configuredFirePremium);
        map.put(RiskType.THEFT, configuredTheftPremium);
        return map;
    }
}

File: src/main/kotlin/org/proofit/premiumcalc/calculation/PremiumFactory.kt
package org.proofit.premiumcalc.calculation

import org.proofit.premiumcalc.exceptions.ApplicationException
import org.proofit.premiumcalc.model.RiskType

object PremiumFactory {
    // TODO: add configurability for calculators
    var configuredFirePremium = FirePremium()
    var configuredTheftPremium = TheftPremium()
    fun getPremium(riskType: RiskType?): Premium {
        return when (riskType) {
            RiskType.FIRE -> configuredFirePremium
            RiskType.THEFT -> configuredTheftPremium
            else -> throw ApplicationException("No calculator for Policy Sub Object defined.")
        }
    }
}
